meson.build

```diff
@@ -699,3 +700,20 @@ if get_option('python_bindings')
     include_directories: [includes],
     dependencies: [cpython] + deps)
 endif
+
+#############################################################################
+## Lc0ctl tool
+#############################################################################
+
+if get_option('lc0ctl')
+  lc0ctl_files = [
+    'src/lc0ctl/describenet.cc',
+    'src/lc0ctl/leela2onnx.cc',
+    'src/lc0ctl/onnx2leela.cc',
+  ]
+
+  executable('lc0ctl', 'src/lc0ctl/lc0ctl_main.cc',
+    files + lc0ctl_files,
```

Comment:
This is a bit wasteful. What about something like:
```diff
diff --git a/meson.build b/meson.build
index 6a9853c..25f0c57 100644
--- a/meson.build
+++ b/meson.build
@@ -633,8 +633,10 @@ if get_option('embed')
   add_project_arguments('-DEMBED', language : 'cpp')
 endif
 
+lc0_lib = static_library('lc0_lib', files, include_directories: includes, dependencies: deps)
+
 executable('lc0', 'src/main.cc',
-  files, include_directories: includes, dependencies: deps, install: true)
+  include_directories: includes, link_with: lc0_lib, install: true)
 
 #############################################################################
 ## Tests
@@ -642,7 +644,6 @@ executable('lc0', 'src/main.cc',
 
 if get_option('gtest')
   gtest = dependency('gtest', fallback: ['gtest', 'gtest_dep'])
-  lc0_lib = library('lc0_lib', files, include_directories: includes, dependencies: deps)
 
   test('ChessBoard',
     executable('chessboard_test', 'src/chess/board_test.cc',
@@ -713,7 +714,7 @@ if get_option('lc0ctl')
   ]
 
   executable('lc0ctl', 'src/lc0ctl/lc0ctl_main.cc',
-    files + lc0ctl_files,
-    include_directories: includes, dependencies: deps, install: true)
+    lc0ctl_files,
+    include_directories: includes, link_with: lc0_lib, install: true)
 
 endif
```

Reply:
This seems to work even without an older `net.pb.h` but I don't like it:
```diff
diff --git a/meson.build b/meson.build
index 6a9853c..6961e24 100644
--- a/meson.build
+++ b/meson.build
@@ -90,12 +90,11 @@ if run_command('scripts/checkdir.py', 'libs/lczero-common/proto').returncode() !
   endif
 endif
 
+files += 'src/utils/protomessage.cc'
 pb_files = [
-  'src/utils/protomessage.cc',
   gen.process('libs/lczero-common/proto/net.proto',
     preserve_path_from : meson.current_source_dir() + '/libs/lczero-common/')
 ]
-files += pb_files
 
 # Extract git short revision.
 short_rev = 'unknown'
@@ -146,7 +145,7 @@ gen_proto_src = generator(compile_proto, output: ['@BASENAME@.pb.h'],
     '--cpp_out=@BUILD_DIR@',
     '@INPUT@'])
 
-files += gen_proto_src.process('src/neural/onnx/onnx.proto',
+pb_files += gen_proto_src.process('src/neural/onnx/onnx.proto',
   preserve_path_from : meson.current_source_dir() + '/src/')
 
 #############################################################################
@@ -633,8 +632,10 @@ if get_option('embed')
   add_project_arguments('-DEMBED', language : 'cpp')
 endif
 
+lc0_lib = static_library('lc0_lib', pb_files, files, include_directories: includes, dependencies: deps)
+
 executable('lc0', 'src/main.cc',
-  files, include_directories: includes, dependencies: deps, install: true)
+  pb_files, include_directories: includes, link_with: lc0_lib, install: true)
 
 #############################################################################
 ## Tests
@@ -642,7 +643,6 @@ executable('lc0', 'src/main.cc',
 
 if get_option('gtest')
   gtest = dependency('gtest', fallback: ['gtest', 'gtest_dep'])
-  lc0_lib = library('lc0_lib', files, include_directories: includes, dependencies: deps)
 
   test('ChessBoard',
     executable('chessboard_test', 'src/chess/board_test.cc',
@@ -713,7 +713,7 @@ if get_option('lc0ctl')
   ]
 
   executable('lc0ctl', 'src/lc0ctl/lc0ctl_main.cc',
-    files + lc0ctl_files,
-    include_directories: includes, dependencies: deps, install: true)
+    pb_files, lc0ctl_files,
+    include_directories: includes, link_with: lc0_lib, install: true)
 
 endif
```
An alternative is:
```diff
diff --git a/meson.build b/meson.build
index 6a9853c..7b3b51f 100644
--- a/meson.build
+++ b/meson.build
@@ -66,11 +66,6 @@ includes += include_directories('third_party', is_system: true)
 
 # Compiling protobufs.
 compile_proto = find_program('scripts/compile_proto.py')
-gen = generator(compile_proto, output: ['@BASENAME@.pb.h'],
-  arguments : [
-    '--proto_path=@CURRENT_SOURCE_DIR@/libs/lczero-common',
-    '--cpp_out=@BUILD_DIR@',
-    '@INPUT@'])
 
 # Handle submodules.
 git = find_program('git', required: false)
@@ -90,12 +85,16 @@ if run_command('scripts/checkdir.py', 'libs/lczero-common/proto').returncode() !
   endif
 endif
 
+files += 'src/utils/protomessage.cc'
 pb_files = [
-  'src/utils/protomessage.cc',
-  gen.process('libs/lczero-common/proto/net.proto',
-    preserve_path_from : meson.current_source_dir() + '/libs/lczero-common/')
+    custom_target(
+    'net.pb.h',
+    output : 'net.pb.h',
+    input : 'libs/lczero-common/proto/net.proto',
+    command : [compile_proto, '--proto_path=@CURRENT_SOURCE_DIR@/libs/lczero-common',
+        '--cpp_out=@BUILD_ROOT@', '@INPUT@']
+    )
 ]
-files += pb_files
 
 # Extract git short revision.
 short_rev = 'unknown'
@@ -140,14 +139,13 @@ elif get_option('malloc') != ''
 endif
 
 # ONNX protobufs.
-gen_proto_src = generator(compile_proto, output: ['@BASENAME@.pb.h'],
-  arguments : [
-    '--proto_path=@CURRENT_SOURCE_DIR@/src',
-    '--cpp_out=@BUILD_DIR@',
-    '@INPUT@'])
-
-files += gen_proto_src.process('src/neural/onnx/onnx.proto',
-  preserve_path_from : meson.current_source_dir() + '/src/')
+pb_files += custom_target(
+    'onnx.pb.h',
+    output : 'onnx.pb.h',
+    input : 'src/neural/onnx/onnx.proto',
+    command : [compile_proto, '--proto_path=@CURRENT_SOURCE_DIR@/src',
+        '--cpp_out=@BUILD_ROOT@', '@INPUT@'],
+)
 
 #############################################################################
 ## Main files
@@ -633,8 +631,10 @@ if get_option('embed')
   add_project_arguments('-DEMBED', language : 'cpp')
 endif
 
+lc0_lib = static_library('lc0_lib', pb_files, files, include_directories: includes, dependencies: deps)
+
 executable('lc0', 'src/main.cc',
-  files, include_directories: includes, dependencies: deps, install: true)
+  pb_files, include_directories: includes, link_with: lc0_lib, install: true)
 
 #############################################################################
 ## Tests
@@ -642,7 +642,6 @@ executable('lc0', 'src/main.cc',
 
 if get_option('gtest')
   gtest = dependency('gtest', fallback: ['gtest', 'gtest_dep'])
-  lc0_lib = library('lc0_lib', files, include_directories: includes, dependencies: deps)
 
   test('ChessBoard',
     executable('chessboard_test', 'src/chess/board_test.cc',
@@ -713,7 +712,7 @@ if get_option('lc0ctl')
   ]
 
   executable('lc0ctl', 'src/lc0ctl/lc0ctl_main.cc',
-    files + lc0ctl_files,
-    include_directories: includes, dependencies: deps, install: true)
+    pb_files, lc0ctl_files,
+    include_directories: includes, link_with: lc0_lib, install: true)
 
 endif
```
Probably too much trouble, just moving all the tools into lc0 is starting to look tempting.

Reply:
Using a library like that messes the backend registration (which in retrospect should have been obvious). I had done this right in #1031...

Reply:
Just merged everything into a single binary as discussed.